x-pack/plugins/watcher/public/sections/watch_edit/components/json_watch_edit_form.tsx

```diff
@@ -24,48 +24,32 @@ import { documentationLinks } from '../../../lib/documentation_links';
 import { onWatchSave, saveWatch } from '../watch_edit_actions';
 import { WatchContext } from './watch_context';
 
-const JSON_WATCH_IDS = {
-  ID: 'watchId',
-  NAME: 'watchName',
-  JSON: 'watchJson',
-};
-
-function validateId(id: string) {
-  const regex = /^[A-Za-z0-9\-\_]+$/;
-  if (!id) {
-    return i18n.translate('xpack.watcher.sections.watchEdit.json.error.requiredIdText', {
-      defaultMessage: 'ID is required',
-    });
-  } else if (!regex.test(id)) {
-    return i18n.translate('xpack.watcher.sections.watchEdit.json.error.invalidIdText', {
-      defaultMessage: 'ID can only contain letters, underscores, dashes, and numbers.',
-    });
-  }
-  return false;
-}
-
 export const JsonWatchEditForm = ({
   urlService,
   licenseService,
-  watchJsonString,
-  setWatchJsonString,
-  errors,
-  setErrors,
-  isShowingErrors,
-  setIsShowingErrors,
 }: {
   urlService: any;
   licenseService: any;
-  watchJsonString: string;
-  setWatchJsonString: (json: string) => void;
-  errors: { [key: string]: string[] };
-  setErrors: (errors: { [key: string]: string[] }) => void;
-  isShowingErrors: boolean;
-  setIsShowingErrors: (isShowingErrors: boolean) => void;
 }) => {
   const { watch, setWatchProperty } = useContext(WatchContext);
   // hooks
-  const [modal, setModal] = useState<{ message: string } | null>(null);
+  const [modal, setModal] = useState<{ title: string; message: string } | null>(null);
+  const { errors } = watch.validate();
+  const hasErrors = !!Object.keys(errors).find(errorKey => errors[errorKey].length >= 1);
+  const prevWatchString = usePrevious(watch.watchString);
+
+  function usePrevious(value: any) {
+    const ref = useRef(null);
+    useEffect(() => {
+      ref.current = value;
+    });
+    return ref.current;
+  }
+
+  if (prevWatchString !== watch.watchString && errors.json.length === 0) {
+    setWatchProperty('watch', JSON.parse(watch.watchString));
```

Comment:
I'm struggling with the best way to validate the watch json and update the watch accordingly. The ace editor requires a json string, and we also can't parse it and update the watch property unless it is valid json. This works, but not sure if there's a better approach.

Reply:
reworked this based on our discussion